File: ChordVoicingsCreator/src/main/java/com/campoy/chord/voicing/creator/model/tinwhistle/SixHoleWhistle.java
package com.campoy.chord.voicing.creator.model.tinwhistle;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import com.campoy.chord.voicing.creator.model.musictheory.Interval;
import com.campoy.chord.voicing.creator.model.musictheory.Note;
import com.campoy.chord.voicing.creator.model.musictheory.Scale;
import com.google.common.collect.Sets;

import lombok.Data;

@Data
public class SixHoleWhistle {

    private final Note baseNote;
    
    private Set<NoteAndFingering> naturalNotes = null;
    private Set<NoteAndFingering> crossFingeredNotes = null;
    private Set<NoteAndFingering> allPossibleNotes = null;
    
    public Set<NoteAndFingering> getCrossFingeredNotes(){
        if(crossFingeredNotes == null) {
            Set<NoteAndFingering> notes = new HashSet<>();
            for( WhistleFingering crossFingering : SixHoleWhistleFingerings.crossFingerings ) {
                notes.add( new NoteAndFingering(baseNote.up(crossFingering.getSemitonesFromRoot()), 
                                                crossFingering));
            }
            crossFingeredNotes = notes;
        }
        return crossFingeredNotes;
    }
    
    public Set<NoteAndFingering> getNaturalNotes(){
        if(naturalNotes == null) {
            Set<NoteAndFingering> notes = new HashSet<>();
            for( WhistleFingering crossFingering : SixHoleWhistleFingerings.naturalFingerings ) {
                notes.add( new NoteAndFingering(baseNote.up(crossFingering.getSemitonesFromRoot()), 
                                                crossFingering));
            }
            naturalNotes = notes;
        }
        return naturalNotes;
    }
    
    public Set<NoteAndFingering> getAllPossibleNotesWithFingerings(){
        if(allPossibleNotes == null) {
            allPossibleNotes = Sets.union(getNaturalNotes(), getCrossFingeredNotes());
        }
        return allPossibleNotes;
    }
    
    public Set<Note> getAllPossibleNotes(){
        return getAllPossibleNotesWithFingerings()
               .stream()
               .map(noteAndFingering -> noteAndFingering.getNote())
               .collect(Collectors.toSet());
    }

    
//    public boolean requiresCrossFingering(Note note) {
//        return getCrossFingeredNotes().contains(note);
//    }
//    
//    public boolean requiresCrossFingering(List<Note> scaleNotes) {
//        for( Note scaleNote : scaleNotes ) {
//            if( requiresCrossFingering(scaleNote) ) {
//                return true;
//            }
//        }
//        return false;
//    }
    
    public List<NoteAndFingering> getNeededCrossFingerings(List<Note> scaleNotes){
        return getCrossFingeredNotes()
                .stream()
                .filter(noteAndFingering -> scaleNotes.contains(noteAndFingering.getNote()))
                .collect(Collectors.toList());
    }
    

}

File: PancakeMusicProjectsTS/src/model/tinwhistle/SixHoleWhistle.ts
import { EqualsTrait } from "../general/EqualsTrait.js";
import { Note } from "../musictheory/Note.js";
import { NoteAndFingering } from "./NoteAndFingering.js";
import { WhistleFingerings } from "./WhistleFingerings.js";

export class SixHoleWhistle implements EqualsTrait{
  private baseNote: Note;
  private naturalNotes: Set<NoteAndFingering> | null = null;
  private crossFingeredNotes: Set<NoteAndFingering> | null = null;
  private allPossibleNotes: Set<NoteAndFingering> | null = null;

  constructor(baseNote: Note) {
    this.baseNote = baseNote;
  }

  equals(other: any): boolean {
    if(!(other instanceof SixHoleWhistle)){
      return false;
    }
    return this.baseNote.equals(other.baseNote);
  }

  public getCrossFingeredNotes(): Set<NoteAndFingering> {
    if (this.crossFingeredNotes === null) {
      const notes = new Set<NoteAndFingering>();
      for (const crossFingering of WhistleFingerings.sixHolesCrossFingerings) {
        notes.add(new NoteAndFingering(this.baseNote.up(crossFingering.getSemitonesFromRoot()), crossFingering));
      }
      this.crossFingeredNotes = notes;
    }
    return this.crossFingeredNotes;
  }

  public getNaturalNotes(): Set<NoteAndFingering> {
    if (this.naturalNotes === null) {
      const notes = new Set<NoteAndFingering>();
      for (const naturalFingering of WhistleFingerings.naturalSixHolesFingerings) {
        notes.add(new NoteAndFingering(this.baseNote.up(naturalFingering.getSemitonesFromRoot()), naturalFingering));
      }
      this.naturalNotes = notes;
    }
    return this.naturalNotes;
  }

  public getAllPossibleNotesWithFingerings(): Set<NoteAndFingering> {
    if (this.allPossibleNotes === null) {
      this.allPossibleNotes = new Set([...this.getNaturalNotes(), ...this.getCrossFingeredNotes()]);
    }
    return this.allPossibleNotes;
  }

  public getAllPossibleNotes(): Set<Note> {
    return new Set([...this.getAllPossibleNotesWithFingerings()].map(noteAndFingering => noteAndFingering.getNote()));
  }

  public getNeededCrossFingerings(scaleNotes: Note[]): NoteAndFingering[] {
    return [...this.getCrossFingeredNotes()].filter(noteAndFingering => scaleNotes.includes(noteAndFingering.getNote()));
  }

  public getFingerings(scaleNotes: Note[]): NoteAndFingering[] {
    return [...this.getAllPossibleNotesWithFingerings()].filter(noteAndFingering => scaleNotes.includes(noteAndFingering.getNote()));
  }
}